Index: R-4.4.0/Makefile.in
===================================================================
--- R-4.4.0.orig/Makefile.in
+++ R-4.4.0/Makefile.in
@@ -91,6 +91,13 @@ javaconf: R
 javaconf: R
 	@$(MAKE) stamp-java
 
+install-wasm:
+	@mkdir -p "$(prefix)/lib/R/doc/html"
+	@for d in $(SUBDIRS); do \
+	  (cd $${d} && $(MAKE) install) \
+	done
+	-@(cd share && $(MAKE) install-zoneinfo)
+	-@(cd src/main && $(MAKE) $@)
 
 install install-strip: installdirs svnonly
 	@for d in $(SUBDIRS); do \
Index: R-4.4.0/src/main/Makefile.in
===================================================================
--- R-4.4.0.orig/src/main/Makefile.in
+++ R-4.4.0/src/main/Makefile.in
@@ -148,7 +148,57 @@ $(R_binary): $(R_bin_OBJECTS) $(R_bin_DEPENDENCIES)
 
 $(R_binary): $(R_bin_OBJECTS) $(R_bin_DEPENDENCIES)
 	$(MAIN_LINK) -o $@ $(R_bin_OBJECTS) $(R_bin_LDADD)
+
+MAIN_WEBR_LDADD  = --use-preload-plugins
+MAIN_WEBR_LDADD += --preload-file "$(prefix)/tmp/lib@/usr/lib"
+MAIN_WEBR_LDADD += --extern-pre-js $(prefix)/pre.js
+
+ifdef WEBR_REPO
+MAIN_WEBR_LDADD += --preload-file "$(WEBR_REPO)@/repo"
+endif
+
+ifdef WEBR_LIB
+MAIN_WEBR_LDADD += --preload-file "${WEBR_LIB}@/usr/lib/R/library"
+endif
+
+$(R_binary).js: $(R_bin_OBJECTS) $(R_bin_DEPENDENCIES)
+	@rm -rf "$(prefix)/tmp" && mkdir -p "$(prefix)/tmp/lib/R/"
+	@cp -a "$(prefix)/lib/R/library" "$(prefix)/tmp/lib/R/"
+	@cp -a "$(prefix)/lib/R/etc" "$(prefix)/tmp/lib/R/"
+	@cp -a "$(prefix)/lib/R/modules" "$(prefix)/tmp/lib/R/"
+	@rm -r "$(prefix)/tmp/lib/R/library/translations"
+	@rm -r "$(prefix)/tmp/lib/R/library/tcltk"
+	@rm -r "$(prefix)/tmp/lib/R/library/parallel"
+	@rm -rf "$(prefix)/tmp/lib/R/library/grDevices/libs/cairo.so"
+	@find "$(prefix)/tmp/lib/R" -type d -name 'afm' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'demo' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'enc' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'doc' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'fonts' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'help' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'html' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'misc' -exec rm -r "{}" +
+	@find "$(prefix)/tmp/lib/R" -type d -name 'tests' -exec rm -r "{}" +
+	$(MAIN_LINK) $(MAIN_WEBR_LDADD) \
+	-o $(prefix)/dist/$(R_binary).js \
+	$(R_bin_OBJECTS) -lRlapack $(FORTRAN_WASM_LDADD) $(R_bin_LDADD)
 
+install-wasm: $(R_bin_DEPENDENCIES)
+	@$(MKINSTALLDIRS) "$(prefix)/dist"
+	@cp "$(Rexeclibdir)/libRblas.so" "$(prefix)/dist/libRblas.so"
+	@cp "$(Rexeclibdir)/libRlapack.so" "$(prefix)/dist/libRlapack.so"
+	@echo "Building virtual filesystem:"
+	@$(LZFS) -v -d $(VFSDIR) -u vfs $(WASM_LAZY_VFS) -o $(prefix)/pre.js
+	@$(MAKE) $(R_binary).js
+# Patch Emscripten library to redirect a warning message to console.warn()
+	@sed -i.bak 's/out("LazyFiles/console.warn("LazyFiles/' $(prefix)/dist/$(R_binary).js
+	@rm $(prefix)/dist/$(R_binary).js.bak
+# Redirect $(WASM)/lib R binaries to $(HOST)/lib R binaries
+	@mv $(prefix)/lib/R/bin/R $(prefix)/lib/R/bin/R.orig
+	@mv $(prefix)/lib/R/bin/Rscript $(prefix)/lib/R/bin/Rscript.orig
+	@ln -s $(R_HOST)/lib/R/bin/R $(prefix)/lib/R/bin/R
+	@ln -s $(R_HOST)/lib/R/bin/Rscript $(prefix)/lib/R/bin/Rscript
+
 libR.a: $(OBJECTS) $(STATIC_LIBS)
 	-@mkdir libs
 	@(cd libs; for l in $(STATIC_LIBS); do $(AR) -x ../$$l; done)
