Index: R-4.4.0/src/modules/lapack/dlapack.f
===================================================================
--- R-4.4.0.orig/src/modules/lapack/dlapack.f
+++ R-4.4.0/src/modules/lapack/dlapack.f
@@ -113014,6 +113014,20 @@
 *> \ingroup pstf2
 *
 *  =====================================================================
+      INTEGER FUNCTION MAXLOCD( N, ARR )
+      INTEGER          I, N
+      DOUBLE PRECISION ARR( N ), TMP
+
+      MAXLOCD = 0
+      TMP = -HUGE(TMP)
+      DO I = 1, N
+         IF( ARR(I) > TMP) THEN
+            MAXLOCD = I
+            TMP = ARR(I)
+         END IF
+      END DO
+      END
+
       SUBROUTINE DPSTF2( UPLO, N, A, LDA, PIV, RANK, TOL, WORK, INFO )
 *
 *  -- LAPACK computational routine --
@@ -113038,7 +113052,7 @@
 *     ..
 *     .. Local Scalars ..
       DOUBLE PRECISION   AJJ, DSTOP, DTEMP
-      INTEGER            I, ITEMP, J, PVT
+      INTEGER            I, ITEMP, J, PVT, MAXLOCD
       LOGICAL            UPPER
 *     ..
 *     .. External Functions ..
@@ -113050,7 +113064,7 @@
       EXTERNAL           DGEMV, DSCAL, DSWAP, XERBLA
 *     ..
 *     .. Intrinsic Functions ..
-      INTRINSIC          MAX, SQRT, MAXLOC
+      INTRINSIC          MAX, SQRT
 *     ..
 *     .. Executable Statements ..
 *
@@ -113131,7 +113145,7 @@
   120       CONTINUE
 *
             IF( J.GT.1 ) THEN
-               ITEMP = MAXLOC( WORK( (N+J):(2*N) ), 1 )
+               ITEMP = MAXLOCD(1+(2*N)-(N+J), WORK((N+J):(2*N)))
                PVT = ITEMP + J - 1
                AJJ = WORK( N+PVT )
                IF( AJJ.LE.DSTOP.OR.DISNAN( AJJ ) ) THEN
@@ -113194,7 +113208,7 @@
   140       CONTINUE
 *
             IF( J.GT.1 ) THEN
-               ITEMP = MAXLOC( WORK( (N+J):(2*N) ), 1 )
+               ITEMP = MAXLOCD(1+(2*N)-(N+J), WORK((N+J):(2*N)))
                PVT = ITEMP + J - 1
                AJJ = WORK( N+PVT )
                IF( AJJ.LE.DSTOP.OR.DISNAN( AJJ ) ) THEN
@@ -113422,7 +113436,7 @@
 *     ..
 *     .. Local Scalars ..
       DOUBLE PRECISION   AJJ, DSTOP, DTEMP
-      INTEGER            I, ITEMP, J, JB, K, NB, PVT
+      INTEGER            I, ITEMP, J, JB, K, NB, PVT, MAXLOCD
       LOGICAL            UPPER
 *     ..
 *     .. External Functions ..
@@ -113435,7 +113449,7 @@
       EXTERNAL           DGEMV, DPSTF2, DSCAL, DSWAP, DSYRK, XERBLA
 *     ..
 *     .. Intrinsic Functions ..
-      INTRINSIC          MAX, MIN, SQRT, MAXLOC
+      INTRINSIC          MAX, MIN, SQRT
 *     ..
 *     .. Executable Statements ..
 *
@@ -113537,7 +113551,7 @@
   120             CONTINUE
 *
                   IF( J.GT.1 ) THEN
-                     ITEMP = MAXLOC( WORK( (N+J):(2*N) ), 1 )
+                     ITEMP = MAXLOCD(1+(2*N)-(N+J), WORK((N+J):(2*N)))
                      PVT = ITEMP + J - 1
                      AJJ = WORK( N+PVT )
                      IF( AJJ.LE.DSTOP.OR.DISNAN( AJJ ) ) THEN
@@ -113624,7 +113638,7 @@
   160             CONTINUE
 *
                   IF( J.GT.1 ) THEN
-                     ITEMP = MAXLOC( WORK( (N+J):(2*N) ), 1 )
+                     ITEMP = MAXLOCD(1+(2*N)-(N+J), WORK((N+J):(2*N)))
                      PVT = ITEMP + J - 1
                      AJJ = WORK( N+PVT )
                      IF( AJJ.LE.DSTOP.OR.DISNAN( AJJ ) ) THEN
