Index: fontconfig-2.12.5/src/fcxml.c
===================================================================
--- fontconfig-2.12.5.orig/src/fcxml.c
+++ fontconfig-2.12.5/src/fcxml.c
@@ -3372,9 +3372,6 @@ FcConfigParseAndLoad (FcConfig	    *config,
 	ret = FcTrue;
 	goto bail0;
     }
-    if (sysroot)
-	filename = FcStrBuildFilename (sysroot, f, NULL);
-    else
 	filename = FcStrdup (f);
     FcStrFree (f);
 
