Index: fontconfig-2.12.5/src/fccache.c
===================================================================
--- fontconfig-2.12.5.orig/src/fccache.c
+++ fontconfig-2.12.5/src/fccache.c
@@ -584,7 +584,7 @@ FcCacheTimeValid (FcConfig *config, FcCache *cache, st
 	printf ("FcCacheTimeValid dir \"%s\" cache checksum %d dir checksum %d\n",
 		FcCacheDir (cache), cache->checksum, (int) dir_stat->st_mtime);
 #endif
-
+    if (strcmp(FcCacheDir(cache), "/usr/share/fonts") == 0) return FcTrue;
     return cache->checksum == (int) dir_stat->st_mtime && fnano;
 }
 
