Index: R-4.3.3/src/library/grDevices/src/grDevices.h
===================================================================
--- R-4.3.3.orig/src/library/grDevices/src/grDevices.h
+++ R-4.3.3/src/library/grDevices/src/grDevices.h
@@ -120,5 +120,5 @@ SEXP devUp(void);
 SEXP setMask(SEXP args);
 SEXP defineGroup(SEXP args);
 SEXP useGroup(SEXP args);
-SEXP devUp(void);
+SEXP devUp(SEXP args);
 SEXP stroke(SEXP args);
Index: R-4.3.3/src/library/grDevices/src/group.c
===================================================================
--- R-4.3.3.orig/src/library/grDevices/src/group.c
+++ R-4.3.3/src/library/grDevices/src/group.c
@@ -65,7 +65,7 @@ SEXP devUp(void) {
     return R_NilValue;
 }
 
-SEXP devUp(void) {
+SEXP devUp(SEXP args) {
     pDevDesc dd = GEcurrentDevice()->dev;
     double left, right, bottom, top;
     SEXP ans;
