Index: R-4.2.2/src/library/stats/src/distn.c
===================================================================
--- R-4.2.2.orig/src/library/stats/src/distn.c
+++ R-4.2.2/src/library/stats/src/distn.c
@@ -460,13 +460,13 @@ extern void wilcox_free(void);
 extern void wilcox_free(void);
 */
 
-SEXP stats_signrank_free(void)
+SEXP stats_signrank_free(SEXP args)
 {
     signrank_free();
     return R_NilValue;
 }
 
-SEXP stats_wilcox_free(void)
+SEXP stats_wilcox_free(SEXP args)
 {
     wilcox_free();
     return R_NilValue;
Index: R-4.2.2/src/library/stats/src/statsR.h
===================================================================
--- R-4.2.2.orig/src/library/stats/src/statsR.h
+++ R-4.2.2/src/library/stats/src/statsR.h
@@ -205,5 +205,5 @@ SEXP Smirnov_sim(SEXP sr, SEXP sc, SEXP sB, SEXP twosi
 
 SEXP Smirnov_sim(SEXP sr, SEXP sc, SEXP sB, SEXP twosided);
 
-SEXP stats_signrank_free(void);
-SEXP stats_wilcox_free(void);
+SEXP stats_signrank_free(SEXP args);
+SEXP stats_wilcox_free(SEXP args);
