Index: R-4.3.0/src/library/grDevices/src/cairo/cairoFns.c
===================================================================
--- R-4.3.0.orig/src/library/grDevices/src/cairo/cairoFns.c
+++ R-4.3.0/src/library/grDevices/src/cairo/cairoFns.c
@@ -1679,10 +1679,13 @@ PangoCairo_Text(double x, double y,
    with freetype2/fontconfig (FT implies FC in Cairo).
 */
 
+#undef TYPEOF
 #include <cairo-ft.h>
 #include <ft2build.h> // currently included by cairo-ft.h
 #include FT_FREETYPE_H
 #include <fontconfig/fontconfig.h>
+#undef TYPEOF
+#define TYPEOF(x)     ((x)->sxpinfo.type)
 
 SEXP in_CairoFT(void) 
 {
