Index: fontconfig-2.17.1/src/fccfg.c
===================================================================
--- fontconfig-2.17.1.orig/src/fccfg.c
+++ fontconfig-2.17.1/src/fccfg.c
@@ -2634,10 +2634,7 @@ FcConfigGetFilename (FcConfig      *conf
 	for (p = path; *p; p++) {
 	    FcChar8 *s;

-	    if (sysroot)
-		s = FcStrBuildFilename (sysroot, *p, NULL);
-	    else
-		s = *p;
+		s = FcStrdup(*p);
 	    file = FcConfigFileExists (s, url);
 	    if (sysroot)
 		FcStrFree (s);
