Index: R-4.2.2/src/unix/sys-unix.c
===================================================================
--- R-4.2.2.orig/src/unix/sys-unix.c
+++ R-4.2.2/src/unix/sys-unix.c
@@ -60,6 +60,19 @@
 
 #include <errno.h>
 
+#ifdef __EMSCRIPTEN__
+/*
+ * The sigsuspend function is currently unsupported in emscripten.
+ * Instead we immediately return -1 and set errno, as if the
+ * suspended process had immediately been delivered a signal.
+ */
+int sigsuspend(const sigset_t *mask)
+{
+    errno = EINVAL;
+    return -1;
+}
+#endif
+
 /*
  *  4) INITIALIZATION AND TERMINATION ACTIONS
  */
