Index: R-4.1.3/src/main/envir.c
===================================================================
--- R-4.1.3.orig/src/main/envir.c
+++ R-4.1.3/src/main/envir.c
@@ -920,6 +920,9 @@ static SEXP findVarLocInFrame(SEXP rho, 
 	frame = FRAME(rho);
 	while (frame != R_NilValue && TAG(frame) != symbol)
 	    frame = CDR(frame);
+#ifdef __EMSCRIPTEN__
+		if((unsigned int) frame > 0x30000000) return R_NilValue;
+#endif
 	return frame;
     }
     else {
@@ -1032,6 +1035,9 @@ SEXP findVarInFrame3(SEXP rho, SEXP symb
 	    if (TAG(frame) == symbol)
 		return BINDING_VALUE(frame);
 	    frame = CDR(frame);
+#ifdef __EMSCRIPTEN__
+         if((unsigned int) frame > 0x30000000) return R_UnboundValue;
+#endif
 	}
     }
     else {
@@ -1081,6 +1087,9 @@ static Rboolean existsVarInFrame(SEXP rh
 	    if (TAG(frame) == symbol)
 		return TRUE;
 	    frame = CDR(frame);
+#ifdef __EMSCRIPTEN__
+         if((unsigned int) frame > 0x30000000) return FALSE;
+#endif
 	}
     }
     else {
@@ -1775,6 +1784,9 @@ static SEXP setVarInFrame(SEXP rho, SEXP
 		return symbol;
 	    }
 	    frame = CDR(frame);
+#ifdef __EMSCRIPTEN__
+         if((unsigned int) frame > 0x30000000) return R_NilValue;
+#endif
 	}
     } else {
 	/* Do the hash table thing */
Index: R-4.1.3/src/main/eval.c
===================================================================
--- R-4.1.3.orig/src/main/eval.c
+++ R-4.1.3/src/main/eval.c
@@ -1708,7 +1708,11 @@ static R_INLINE void R_CleanupEnvir(SEXP
 	    refs -= countCycleRefs(rho, val);
 	if (refs == 0) {
 	    for (SEXP b = FRAME(rho);
-		 b != R_NilValue && REFCNT(b) == 1;
+        b != R_NilValue && REFCNT(b) == 1
+#ifdef __EMSCRIPTEN__
+		&& ((unsigned int)CDR(b))<0x30000000
+#endif
+		;
 		 b = CDR(b)) {
 		if (BNDCELL_TAG(b)) continue;
 		SEXP v = CAR(b);
