Index: R-4.1.3_quilt/src/library/grDevices/NAMESPACE
===================================================================
--- R-4.1.3_quilt.orig/src/library/grDevices/NAMESPACE
+++ R-4.1.3_quilt/src/library/grDevices/NAMESPACE
@@ -71,6 +71,6 @@ if(tools:::.OStype() == "windows") {
     S3method(print, SavedPlots)
     importFrom("utils", "shortPathName")
 } else {
-    export(X11.options, X11Font, X11Fonts,
+    export(canvas, X11.options, X11Font, X11Fonts,
            quartz, quartz.options, quartz.save, quartzFont, quartzFonts)
 }
Index: R-4.1.3_quilt/src/library/grDevices/R/canvas.R
===================================================================
--- /dev/null
+++ R-4.1.3_quilt/src/library/grDevices/R/canvas.R
@@ -0,0 +1,3 @@
+canvas <- function(sel="#plot-canvas", bg="transparent") {
+       .External("canvas_new_device", sel, bg)
+}
Index: R-4.1.3_quilt/src/library/grDevices/src/devCanvas.c
===================================================================
--- /dev/null
+++ R-4.1.3_quilt/src/library/grDevices/src/devCanvas.c
@@ -0,0 +1,501 @@
+/*
+R graphics device targeting the HTML canvas element
+Author: George Stagg, based on the canvas package created by Jeffrey Horner (http://www.rforge.net/canvas/)
+Description: R graphics device interfacing with a HTML canvas element for use with webR.
+License: GPL version 2
+*/
+
+#include <stdio.h>
+
+#include <R.h>
+#include <Rversion.h>
+#include <Rinternals.h>
+#include <R_ext/Rdynload.h>
+#include <R_ext/GraphicsEngine.h>
+#include <R_ext/GraphicsDevice.h>
+
+#if R_VERSION >= R_Version(2,8,0)
+#ifndef NewDevDesc
+#define NewDevDesc DevDesc
+#endif
+#endif
+
+#define CREDC(C) (((unsigned int)(C))&0xff)
+#define CGREENC(C) ((((unsigned int)(C))&0xff00)>>8)
+#define CBLUEC(C) ((((unsigned int)(C))&0xff0000)>>16)
+#define CALPHA(C) ((((unsigned int)(C))&0xff000000)>>24)
+
+#ifdef __EMSCRIPTEN__
+#include <emscripten.h>
+
+/* Interact with the HTML canvas 2D context API by building Javascript
+ * commands and running them with emscripten_run_script().
+ * We avoid using EM_ASM and friends because this code will be compiled
+ * in an emscripten SIDE_MODULE, where they not work at time of writing.
+*/
+#define canvasExecBufferMax 1024
+#define _E(...) "document.querySelector(UTF8ToString(%p)).getContext('2d')." __VA_ARGS__ ";"
+
+#define canvasContextRender(cGD, op, ...) { \
+	const char template[] = _E op; \
+	int ret = snprintf(cGD->buffer, canvasExecBufferMax, template, cGD->sel, ##  __VA_ARGS__); \
+	if (ret < 0 || ret >= canvasExecBufferMax) \
+		error("problem writing canvas context property in canvas graphics device"); \
+	}
+
+#define canvasContextExec(cGD, op, ...) \
+	canvasContextRender(cGD, op, ## __VA_ARGS__) \
+	emscripten_run_script(cGD->buffer)
+
+#define canvasContextExecInt(res, cGD, op, ...) \
+	canvasContextRender(cGD, op, ## __VA_ARGS__) \
+	res = emscripten_run_script_int(cGD->buffer)
+
+#define canvasColor(sel, prop, col) { \
+	if (CALPHA(col)==255) { \
+		canvasContextExec(cGD, (#prop " = 'rgb(%d, %d, %d)'"), CREDC(col), CGREENC(col), CBLUEC(col)); \
+	} else { \
+		canvasContextExec(cGD, (#prop " = 'rgba(%d, %d, %d, %f)'"), CREDC(col), CGREENC(col), CBLUEC(col), ((double)CALPHA(col))/255.); \
+	} \
+}
+
+typedef struct _canvasDesc {
+	/* device specific stuff */
+	int col;
+	int fill;
+
+	/* Line characteristics */
+	double lwd;
+	int lty;
+	R_GE_lineend lend;
+	R_GE_linejoin ljoin;
+	double lmitre;
+
+	char sel[128];
+	char buffer[canvasExecBufferMax];
+	pGEDevDesc RGE;
+} canvasDesc;
+
+void canvasSetLineType( canvasDesc *cGD, pGEcontext gc)
+{
+#ifdef CANVASDEBUG
+	Rprintf("canvasSetLineType()\n");
+#endif
+	/* Line width */
+	if (cGD->lwd != gc->lwd){
+		cGD->lwd = gc->lwd;
+		canvasContextExec(cGD, ("lineWidth = %f"), 2*cGD->lwd);
+	}
+
+	/* Line end: par lend  */
+	if (cGD->lend != gc->lend){
+		cGD->lend = gc->lend;
+		if (cGD->lend == GE_ROUND_CAP)
+			canvasContextExec(cGD, ("lineCap = 'round'"));
+		if (cGD->lend == GE_BUTT_CAP)
+			canvasContextExec(cGD, ("lineCap = 'butt'"));
+		if (cGD->lend == GE_SQUARE_CAP)
+			canvasContextExec(cGD, ("lineCap = 'butt'"));
+	}
+
+	/* Line join: par ljoin */
+	if (cGD->ljoin != gc->ljoin){
+		cGD->ljoin = gc->ljoin;
+		if (cGD->ljoin == GE_ROUND_JOIN)
+			canvasContextExec(cGD, ("lineJoin = 'round'"));
+		if (cGD->ljoin == GE_MITRE_JOIN)
+			canvasContextExec(cGD, ("lineJoin = 'miter'"));
+		if (cGD->ljoin == GE_BEVEL_JOIN)
+			canvasContextExec(cGD, ("lineJoin = 'bevel'"));
+	}
+
+	/* Miter limit */
+	if (cGD->lmitre != gc->lmitre){
+		cGD->lmitre = gc->lmitre;
+		canvasContextExec(cGD, ("miterLimit = %f"), cGD->lmitre);
+	}
+}
+
+void canvasActivate(const pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("canvasActivate()\n");
+#endif
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+}
+
+void canvasCircle(double x, double y, double r, const pGEcontext gc, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("canvasCircle(%f, %f, %f)\n", x, y, r);
+#endif
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+
+	canvasContextExec(cGD, ("beginPath()"));
+	canvasContextExec(cGD, ("arc(%f,%f,%f,0,Math.PI*2,true)"), 2*x, 2*y, 2*r);
+	if (CALPHA(gc->fill)){
+		canvasColor(cGD->sel,fillStyle,gc->fill);
+		canvasContextExec(cGD, ("fill()"));
+	}
+	if (CALPHA(gc->col) && gc->lty!=-1){
+		canvasSetLineType(cGD,gc);
+		canvasColor(cGD->sel,strokeStyle,gc->col);
+		canvasContextExec(cGD, ("stroke()"));
+	}
+}
+
+void canvasClip(double x0, double x1, double y0, double y1, pDevDesc RGD)
+{
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+#ifdef CANVASDEBUG
+	Rprintf("Clip(x0=%f,y0=%f,x1=%f,y1=%f,RGD=0x%x)\n",2.*x0,2.*y0,2.*x1,2.*y1,RGD);
+#endif
+}
+
+void canvasClose(pDevDesc RGD)
+{
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+
+	/* Save plot */
+	free(cGD);
+	RGD->deviceSpecific = NULL;
+#ifdef CANVASDEBUG
+	Rprintf("Close(RGD=0x%x)\n",RGD);
+#endif
+}
+
+void canvasDeactivate(pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("Deactivate(RGD=0x%x)\n",RGD);
+#endif
+}
+static Rboolean canvasLocator(double *x, double *y, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("Locator(x=%f,y=%f,RGD=0x%x)\n",x,y,RGD);
+#endif
+	return FALSE;
+}
+
+void canvasLine(double x1, double y1, double x2, double y2, const pGEcontext gc, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("canvasLine()\n");
+#endif
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+
+	if (CALPHA(gc->col) && gc->lty!=-1){
+	canvasSetLineType(cGD,gc);
+	canvasColor(cGD->sel,strokeStyle,gc->col);
+	canvasContextExec(cGD, ("beginPath()"));
+	canvasContextExec(cGD, ("moveTo(%f,%f)"), 2*x1, 2*y1);
+	canvasContextExec(cGD, ("lineTo(%f,%f)"), 2*x2, 2*y2);
+	canvasContextExec(cGD, ("stroke()"));
+	}
+}
+
+void canvasMetricInfo(int c, const pGEcontext gc, double* ascent, double* descent, double* width, pDevDesc RGD)
+{
+	/* Unsure if we'll be able to provide this, as this relies entirely on the fonts
+	 * installed on the browser system
+	 */
+	*ascent = *descent = *width = 0.0;
+#ifdef CANVASDEBUG
+	Rprintf("MetricInfo(c=%d,gc=0x%x,ascent=%f,descent=%f,width=%f,RGD=0x%x)\n",c,gc,*ascent,*descent,*width,RGD);
+#endif
+}
+
+void canvasMode(int mode, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("Mode(mode=%d,RGD=0x%x)\n",mode,RGD);
+#endif
+}
+
+void canvasNewPage(const pGEcontext gc, pDevDesc RGD)
+{
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+
+	canvasContextExec(cGD, ("clearRect(0,0,%f,%f)"), 2*RGD->right,2*RGD->bottom);
+
+	/* Set background only if we have a color */
+	if (CALPHA(gc->fill)){
+		canvasColor(cGD->sel,fillStyle,gc->fill);
+		canvasContextExec(cGD, ("fillRect(0,0,%f,%f)"), 2*RGD->right,2*RGD->bottom);
+	}
+
+#ifdef CANVASDEBUG
+	Rprintf("NewPage()\n");
+#endif
+}
+
+void canvasPolygon(int n, double *x, double *y, const pGEcontext gc, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("canvasPolygon()\n");
+#endif
+	int i=1;
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+
+	if(n<2) return;
+
+	canvasSetLineType(cGD,gc);
+
+	canvasContextExec(cGD, ("beginPath()"));
+	canvasContextExec(cGD, ("moveTo(%f,%f)"), 2*x[0], 2*y[0]);
+	while (i<n) { canvasContextExec(cGD, ("lineTo(%f, %f)"), 2*x[i], 2*y[i]); i++; }
+	canvasContextExec(cGD, ("closePath()"));
+	if (CALPHA(gc->fill)) {
+		canvasColor(cGD->sel,fillStyle,gc->fill);
+		canvasContextExec(cGD, ("fill()"));
+	}
+	if (CALPHA(gc->col) && gc->lty!=-1) {
+		canvasColor(cGD->sel,strokeStyle,gc->col);
+		canvasContextExec(cGD, ("stroke()"));
+	}
+}
+
+void canvasPolyline(int n, double *x, double *y, const pGEcontext gc, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("canvasPolyline()\n");
+#endif
+	int i=1;
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+
+	if (n<2) return;
+
+	if (CALPHA(gc->col) && gc->lty!=-1) {
+	canvasContextExec(cGD, ("beginPath()"));
+	canvasContextExec(cGD, ("moveTo(%f,%f)"), 2*x[0], 2*y[0]);
+		while(i<n) { canvasContextExec(cGD, ("lineTo(%f, %f)"), 2*x[i], 2*y[i]); i++; }
+		canvasSetLineType(cGD,gc);
+		canvasColor(cGD->sel,strokeStyle,gc->col);
+		canvasContextExec(cGD, ("stroke()"));
+	}
+}
+
+void canvasRect(double x0, double y0, double x1, double y1, const pGEcontext gc, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+	Rprintf("canvasRect()\n");
+#endif
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+	if (CALPHA(gc->fill)){
+		canvasColor(cGD->sel,fillStyle,gc->fill);
+		canvasContextExec(cGD, ("fillRect(%f,%f,%f,%f)"), 2*x0, 2*y0, 2*x1-2*x0, 2*y1-2*y0);
+	}
+	if (CALPHA(gc->col) && gc->lty!=-1){
+		canvasSetLineType(cGD,gc);
+		canvasColor(cGD->sel,strokeStyle,gc->col);
+		canvasContextExec(cGD, ("strokeRect(%f,%f,%f,%f)"), 2*x0, 2*y0, 2*x1-2*x0, 2*y1-2*y0);
+	}
+}
+
+void canvasSize(double *left, double *right, double *bottom, double *top, pDevDesc RGD)
+{
+	*left = *top = 0.0;
+	*right = RGD->right;
+	*bottom = RGD->bottom;
+#ifdef CANVASDEBUG
+	Rprintf("Size(left=%f,right=%f,bottom=%f,top=%f,RGD=0x%x)\n",*left,*right,*bottom,*top,RGD);
+#endif
+}
+
+static double canvasStrWidth(const char *str, const pGEcontext gc, pDevDesc RGD)
+{
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+	canvasContextExec(cGD, ("font = %f+'px sans-serif'"), 2*gc->ps);
+	/* Use fixed point arithmetic to retrieve results via an int type.
+	 * Ideally, we'd rather use something like
+	 *  double w = EM_ASM_DOUBLE({
+	 *   return document.querySelector(UTF8ToString($0)).getContext('2d').measureText(UTF8ToString($1)).width;
+	 * }, cGD->sel, str);
+	 */
+	int wi = 0;
+	canvasContextExecInt(wi, cGD, ("measureText(UTF8ToString(%p)).width * 10000"), str);
+	return (double)wi/10000.;
+
+#ifdef CANVASDEBUG
+	Rprintf("StrWidth(str=%s,gc=0x%x,RGD=0x%x)\n",str,gc,RGD);
+#endif
+}
+
+void canvasText(double x, double y, const char *str, double rot, double hadj, const pGEcontext gc, pDevDesc RGD)
+{
+#ifdef CANVASDEBUG
+        Rprintf("canvasText()\n");
+#endif
+	canvasDesc *cGD = (canvasDesc *)RGD->deviceSpecific;
+	canvasContextExec(cGD, ("font = %f+'px sans-serif'"), 2*gc->ps);
+
+	int wi = 0;
+	if (hadj!=0. || rot != 0.){
+		if (rot!=0.){
+			canvasContextExec(cGD, ("save()"));
+			canvasColor(cGD->sel,fillStyle,gc->col);
+			canvasContextExec(cGD, ("translate(%f,%f)"), 2*x, 2*y);
+			canvasContextExec(cGD, ("rotate(-%f / 180 * Math.PI)"), rot);
+			canvasContextExecInt(wi, cGD, ("measureText(UTF8ToString(%p)).width * 10000"), str);
+			canvasContextExec(cGD, ("fillText(UTF8ToString(%p),-%f*%f,0)"), str, (double)wi/10000., hadj);
+			canvasContextExec(cGD, ("restore()"));
+		} else {
+			canvasColor(cGD->sel,fillStyle,gc->col);
+			canvasContextExecInt(wi, cGD, ("measureText(UTF8ToString(%p)).width * 10000"), str);
+			canvasContextExec(cGD, ("fillText(UTF8ToString(%p),%f-%f*%f,%f)"), str, 2*x, (double)wi/10000., hadj, 2*y);
+		}
+	} else {
+		canvasColor(cGD->sel,fillStyle,gc->col);
+		canvasContextExec(cGD, ("fillText(UTF8ToString(%p),%f,%f)"), str, 2*x, 2*y);
+	}
+}
+
+SEXP void_setPattern(SEXP pattern, pDevDesc RGD) {
+	return R_NilValue;
+}
+void void_releasePattern(SEXP ref, pDevDesc RGD) {
+	return;
+}
+SEXP void_setClipPath(SEXP path, SEXP ref, pDevDesc RGD) {
+	return R_NilValue;
+}
+void void_releaseClipPath(SEXP ref, pDevDesc RGD) {
+	return;
+}
+SEXP void_setMask(SEXP path, SEXP ref, pDevDesc RGD) {
+	return R_NilValue;
+}
+void void_releaseMask(SEXP ref, pDevDesc RGD) {
+	return;
+}
+void void_raster(unsigned int *raster, int w, int h,
+                double x, double y,
+                double width, double height,
+                double rot,
+                Rboolean interpolate,
+                const pGEcontext gc, pDevDesc dd) {
+	return;
+}
+void void_path(double *x, double *y,
+              int npoly, int *nper,
+              Rboolean winding,
+              const pGEcontext gc, pDevDesc dd) {
+	return;
+}
+
+SEXP _canvas_new_device(SEXP args)
+{
+        /* R Graphics Device: in GraphicsDevice.h */
+        pDevDesc RGD;
+
+        /* R Graphics Engine: in GraphicsEngine.h */
+        pGEDevDesc RGE;
+
+        /* canvas Graphics Device */
+        canvasDesc *cGD;
+
+        const char *sel = NULL;
+        int bgcolor;
+
+        SEXP v;
+        args=CDR(args);
+        v=CAR(args); args=CDR(args);
+        if (isString(v)){
+                PROTECT(v);
+                sel = CHAR(STRING_ELT(v,0));
+                UNPROTECT(1);
+        } else {
+                error("selector must be a string");
+        }
+
+        v=CAR(args); args=CDR(args);
+        if (!isString(v) && !isInteger(v) && !isLogical(v) && !isReal(v))
+                error("invalid color specification for `bg'");
+        bgcolor = RGBpar(v, 0);
+#ifdef CANVASDEBUG
+        Rprintf("canvas_new_device()\n");
+#endif
+
+	R_CheckDeviceAvailable();
+
+        if (!(RGD = (pDevDesc)calloc(1, sizeof(NewDevDesc)))){
+		error("calloc failed for canvas device");
+        }
+
+	if (!(cGD = (canvasDesc *)calloc(1, sizeof(canvasDesc)))){
+		free(RGD);
+		error("calloc failed for canvas device");
+        }
+
+        strncpy(cGD->sel, sel, 128);
+
+	RGD->deviceSpecific = (void *) cGD;
+
+	/* Callbacks */
+	RGD->close = canvasClose;
+	RGD->activate = canvasActivate;
+	RGD->deactivate = canvasDeactivate;
+	RGD->size = canvasSize;
+	RGD->newPage = canvasNewPage;
+	RGD->clip = canvasClip;
+	RGD->strWidth = canvasStrWidth;
+	RGD->text = canvasText;
+	RGD->rect = canvasRect;
+	RGD->circle = canvasCircle;
+	RGD->line = canvasLine;
+	RGD->polyline = canvasPolyline;
+	RGD->polygon = canvasPolygon;
+	RGD->locator = canvasLocator;
+	RGD->mode = canvasMode;
+	RGD->metricInfo = canvasMetricInfo;
+	RGD->hasTextUTF8 = TRUE;
+	RGD->strWidthUTF8 = canvasStrWidth;
+	RGD->textUTF8 = canvasText;
+	RGD->wantSymbolUTF8 = TRUE;
+	RGD->path = void_path;
+	RGD->raster = void_raster;
+#if R_GE_version >= 13
+	RGD->setPattern      = void_setPattern;
+	RGD->releasePattern  = void_releasePattern;
+	RGD->setClipPath     = void_setClipPath;
+	RGD->releaseClipPath = void_releaseClipPath;
+	RGD->setMask         = void_setMask;
+	RGD->releaseMask     = void_releaseMask;
+	RGD->deviceVersion = R_GE_definitions;
+#endif
+	/* Initialise RGD */
+	RGD->left = RGD->clipLeft = 0;
+	RGD->top = RGD->clipTop = 0;
+	RGD->right = RGD->clipRight = 504;
+	RGD->bottom = RGD->clipBottom = 504;
+	RGD->xCharOffset = 0.4900;
+	RGD->yCharOffset = 0.3333;
+	RGD->yLineBias = 0.1;
+	RGD->ipr[0] = 1.0/72.0;
+	RGD->ipr[1] = 1.0/72.0;
+	RGD->cra[0] = 0.9 * 10;
+	RGD->cra[1] = 1.2 * 10;
+	RGD->gamma = 1.0;
+	RGD->canClip = FALSE;
+	RGD->canChangeGamma = FALSE;
+	RGD->canHAdj = 2;
+	RGD->startps = 12.0;
+	RGD->startcol = R_RGB(0,0,0);
+	RGD->startfill = bgcolor;
+	RGD->startlty = LTY_SOLID;
+	RGD->startfont = 1;
+	RGD->startgamma = RGD->gamma;
+	RGD->displayListOn = FALSE;
+
+	/* Add to the device list */
+	RGE = GEcreateDevDesc(RGD);
+	cGD->RGE = RGE;
+	GEaddDevice(RGE);
+	GEinitDisplayList(RGE);
+
+	/*return ScalarInteger(1 + GEdeviceNumber(RGE));*/
+	return R_NilValue;
+}
+#endif
+
Index: R-4.1.3_quilt/src/library/grDevices/src/grDevices.h
===================================================================
--- R-4.1.3_quilt.orig/src/library/grDevices/src/grDevices.h
+++ R-4.1.3_quilt/src/library/grDevices/src/grDevices.h
@@ -33,6 +33,9 @@ SEXP R_CreateAtVector(SEXP axp, SEXP usr
 SEXP R_GAxisPars(SEXP usr, SEXP is_log, SEXP nintLog);
 
 SEXP PicTeX(SEXP);
+#ifdef __EMSCRIPTEN__
+extern SEXP _canvas_new_device(SEXP);
+#endif
 
 SEXP PostScript(SEXP);
 SEXP XFig(SEXP);
Index: R-4.1.3_quilt/src/library/grDevices/src/init.c
===================================================================
--- R-4.1.3_quilt.orig/src/library/grDevices/src/init.c
+++ R-4.1.3_quilt/src/library/grDevices/src/init.c
@@ -87,8 +87,14 @@ static const R_CallMethodDef CallEntries
 };
 
 #define EXTDEF(name, n)  {#name, (DL_FUNC) &name, n}
+#ifdef __EMSCRIPTEN__
+SEXP canvas_new_device(SEXP args){return _canvas_new_device(args);}
+#endif
 
 static const R_ExternalMethodDef ExtEntries[] = {
+#ifdef __EMSCRIPTEN__
+    EXTDEF(canvas_new_device, 2),
+#endif
     EXTDEF(PicTeX, 6),
     EXTDEF(PostScript, 19),
     EXTDEF(XFig, 14),
Index: R-4.1.3_quilt/src/library/grDevices/src/Makefile.in
===================================================================
--- R-4.1.3_quilt.orig/src/library/grDevices/src/Makefile.in
+++ R-4.1.3_quilt/src/library/grDevices/src/Makefile.in
@@ -15,7 +15,7 @@ R_INCLUDE_DIR = $(R_HOME)/include
 
 SOURCES_C = axis_scales.c chull.c devices.c init.c stubs.c colors.c \
   clippath.c patterns.c mask.c \
-  devCairo.c devPicTeX.c devPS.c \
+  devCairo.c devPicTeX.c devPS.c devCanvas.c \
   devQuartz.c @BUILD_AQUA_TRUE@ qdBitmap.c qdPDF.c
 SOURCES_M = @BUILD_AQUA_TRUE@ qdCocoa.m
 
