Index: R-4.1.3_quilt/src/main/platform.c
===================================================================
--- R-4.1.3_quilt.orig/src/main/platform.c
+++ R-4.1.3_quilt/src/main/platform.c
@@ -3377,11 +3377,16 @@ do_eSoftVersion(SEXP call, SEXP op, SEXP
        and heuristics, and depends on implementation details
        of dynamic linkers.
     */
+
+#ifdef __EMSCRIPTEN__
+    char *dgemm_name = "_dgemm_";
+#else
 #ifdef HAVE_F77_UNDERSCORE
     char *dgemm_name = "dgemm_";
 #else
     char *dgemm_name = "dgemm";
 #endif
+#endif
 
     Rboolean ok = TRUE;
 
