Index: R-4.1.3/src/main/Rdynload.c
===================================================================
--- R-4.1.3.orig/src/main/Rdynload.c
+++ R-4.1.3/src/main/Rdynload.c
@@ -960,8 +960,8 @@ int R_moduleCdynload(const char *module, int local, in
 
     if(!p) return 0;
 #ifdef R_ARCH
-    snprintf(dllpath, PATH_MAX, "%s%smodules%s%s%s%s%s", p, FILESEP, FILESEP,
-	     R_ARCH, FILESEP, module, SHLIB_EXT);
+    snprintf(dllpath, PATH_MAX, "%s%smodules%s%s%s%s", p, FILESEP, R_ARCH,
+	     FILESEP, module, SHLIB_EXT);
 #else
     snprintf(dllpath, PATH_MAX, "%s%smodules%s%s%s", p, FILESEP, FILESEP,
 	     module, SHLIB_EXT);
