Index: R-4.1.3/src/main/platform.c
===================================================================
--- R-4.1.3.orig/src/main/platform.c
+++ R-4.1.3/src/main/platform.c
@@ -3377,10 +3377,15 @@ do_eSoftVersion(SEXP call, SEXP op, SEXP args, SEXP rh
        and heuristics, and depends on implementation details
        of dynamic linkers.
     */
+
+#ifdef __EMSCRIPTEN__
+    char *dgemm_name = "_dgemm_";
+#else
 #ifdef HAVE_F77_UNDERSCORE
     char *dgemm_name = "dgemm_";
 #else
     char *dgemm_name = "dgemm";
+#endif
 #endif
 
     Rboolean ok = TRUE;
