Index: R-4.1.3/src/library/stats/src/distn.c
===================================================================
--- R-4.1.3.orig/src/library/stats/src/distn.c
+++ R-4.1.3/src/library/stats/src/distn.c
@@ -458,13 +458,13 @@ extern void wilcox_free(void);
 extern void signrank_free(void);
 extern void wilcox_free(void);
 
-SEXP stats_signrank_free(void)
+SEXP stats_signrank_free(SEXP args)
 {
     signrank_free();
     return R_NilValue;
 }
 
-SEXP stats_wilcox_free(void)
+SEXP stats_wilcox_free(SEXP args)
 {
     wilcox_free();
     return R_NilValue;
Index: R-4.1.3/src/library/stats/src/statsR.h
===================================================================
--- R-4.1.3.orig/src/library/stats/src/statsR.h
+++ R-4.1.3/src/library/stats/src/statsR.h
@@ -203,5 +203,5 @@ SEXP d2x2xk(SEXP sK, SEXP sm, SEXP sn, SEXP st, SEXP s
 SEXP chisq_sim(SEXP sr, SEXP sc, SEXP sB, SEXP E);
 SEXP d2x2xk(SEXP sK, SEXP sm, SEXP sn, SEXP st, SEXP srn);
 
-SEXP stats_signrank_free(void);
-SEXP stats_wilcox_free(void);
+SEXP stats_signrank_free(SEXP args);
+SEXP stats_wilcox_free(SEXP args);
