Index: R-4.5.0/src/scripts/R.sh.in
===================================================================
--- R-4.5.0.orig/src/scripts/R.sh.in
+++ R-4.5.0/src/scripts/R.sh.in
@@ -26,10 +26,6 @@ if test "${R_HOME_DIR}" = "@prefix@/@LIB
   esac
 fi
 
-if test -n "${R_HOME}" && \
-   test "${R_HOME}" != "${R_HOME_DIR}"; then
-  echo "WARNING: ignoring environment value of R_HOME"
-fi
 R_HOME="${R_HOME_DIR}"
 export R_HOME
 R_SHARE_DIR="${R_HOME_DIR}/share"
