Index: fontconfig-2.15.0/src/fccfg.c
===================================================================
--- fontconfig-2.15.0.orig/src/fccfg.c
+++ fontconfig-2.15.0/src/fccfg.c
@@ -2716,10 +2716,7 @@ FcConfigGetFilename (FcConfig      *conf
 	{
 	    FcChar8 *s;
 
-	    if (sysroot)
-		s = FcStrBuildFilename (sysroot, *p, NULL);
-	    else
-		s = *p;
+		s = FcStrdup(*p);
 	    file = FcConfigFileExists (s, url);
 	    if (sysroot)
 		FcStrFree (s);
