Index: R-4.4.2/src/main/connections.c
===================================================================
--- R-4.4.2.orig/src/main/connections.c
+++ R-4.4.2/src/main/connections.c
@@ -5356,7 +5356,7 @@ attribute_hidden SEXP do_sink(SEXP call, SEXP op, SEXP
 	    error(_("sink stack is full"));
 	switch_or_tee_stdout(icon, closeOnExit, tee);
     } else {
-	if(icon < 0) {
+	if(icon < 0 || icon == 2) {
 	    R_ReleaseObject(getConnection(R_ErrorCon)->ex_ptr);
 	    R_ErrorCon = 2;
 	} else {
